compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/optimizer/BIRRecordValueOptimizer.java

```diff
@@ -0,0 +1,217 @@
+            }
+        }
+        if (!fpRemoved) {
+            newBBs.add(basicBlock);
+        } else {
+            lastBB.instructions.addAll(basicBlock.instructions);
+        }
+
+        if (basicBlock.terminator.kind == FP_CALL) {
+            handleFPCall(basicBlock);
+        } else if (fpRemoved && valueCreated) {
+            resetBasicBlock(basicBlock);
+        }
+    }
+
+    private void handleFPCall(BIRNode.BIRBasicBlock basicBlock) {
+        BIRTerminator.FPCall fpCall = (BIRTerminator.FPCall) basicBlock.terminator;
+        BIROperand recOperand = recordOperandList.isEmpty() ? null : recordOperandList.get(recordOperandList.size() - 1);
+        BRecordType recordType = recordOperandTypeMap.get(recOperand);
+
+        if (recordType == null || recordType.tsymbol == null) {
+            resetBasicBlock(basicBlock);
+            return;
+        }
+
+        if (!fpCall.fp.variableDcl.name.value.contains(recordType.tsymbol.name.value)) {
+            resetBasicBlock(basicBlock);
+            return;
+        }
+
+        BIRNode.BIRFunction defaultFunction = getDefaultBIRFunction(fpCall.fp.variableDcl.name.value);
+        if (defaultFunction == null) {
+            resetBasicBlock(basicBlock);
+            return;
+        }
+
+        BIRNode.BIRBasicBlock firstBB = defaultFunction.basicBlocks.get(0);
+        lastBB = lastBB != null ? lastBB : basicBlock;
+
+        if (containsOnlyConstantLoad(defaultFunction)) {
+            moveConstLoadInstruction(fpCall, firstBB);
+            lastBB.terminator = null;
+            fpRemoved = true;
+        } else {
+            resetBasicBlock(basicBlock);
+        }
+    }
+
+    private void handleNewStructure(BIRNonTerminator.NewStructure inst) {
+        recordOperandList.remove(inst.rhsOp);
+        valueCreated = true;
+    }
+
+    private void handleNewTypeDesc(BIRNonTerminator inst) {
+        BType referredType = Types.getReferredType(((BIRNonTerminator.NewTypeDesc) inst).type);
+        if (referredType.tag == TypeTags.RECORD) {
+            recordOperandList.add(inst.lhsOp);
+            recordOperandTypeMap.put(inst.lhsOp, (BRecordType) referredType);
+        }
+    }
+
+    private void moveConstLoadInstruction(BIRTerminator.FPCall fpCall, BIRNode.BIRBasicBlock firstBB) {
+
+        BIRNonTerminator.ConstantLoad constantLoad = (BIRNonTerminator.ConstantLoad) firstBB.instructions.get(0);
+        if (firstBB.instructions.size() == 2) {
+            BIRNonTerminator.TypeCast typeCast = (BIRNonTerminator.TypeCast) firstBB.instructions.get(1);
+            String tempVarName = "%temp_" + typeCast.rhsOp.variableDcl.name.value;
+            BIRNode.BIRVariableDcl tempVar;
+            if (typecastVars.containsKey(tempVarName)) {
+                tempVar = typecastVars.get(tempVarName);
+            } else {
+                tempVar = new BIRNode.BIRVariableDcl(null, constantLoad.type, new Name(tempVarName), VarScope.FUNCTION, VarKind.TEMP, "");
```

Comment:
I think we can pass `null` to `metaVarName`. Shall we do it here and in the other places instead of passing `""`